### PR TITLE
[RFC][move-package] add the concept of a move package

### PR DESCRIPTION
**Pull request by mengxu-fb**
## Motivation

Experimental code on grouping several modules together and carry them over in the codebase. Also (hopefully) in preparation when `move-stdlib` is separated from the `diem-framework` and both become independent Move packages.

A `CompiledPackage` maintains the consistency of the raw bytes and the `CompiledModule`, and can be created in two ways:
- `from_bytes` to load existing, on-disk `CompiledModule` (and subject to deserialization and bound checks).
- `from_compilation` to load freshly compiled modules.

A `VerifiedPackage` can only be created by destructing a `CompiledPackage` and verify all its modules against the given dependencies, i.e., following the same flow of bytecode verifier checks.

NOTE: I hate to create a new Cargo.toml file for this simple code but I failed to find a proper location in the `language/` directory...

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

Not quite needed as of now.

## Related PRs

(If this PR adds or changes functionality, please take some time to update the docs at https://github.com/diem/diem/tree/master/developers.diem.com, and link to your PR here.)

### CODE REVIEW

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A move-package 0.1.0 (workspace, path 'language/move-package')
    * features: default
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A move-package 0.1.0 (workspace, path 'language/move-package')
    * features: default
```

File: language/diem-framework/compiled/Cargo.toml [hunk 1]
```diff
@@ -11,15 +11,13 @@ publish = false
...
 [dependencies]
 anyhow = "1.0.38"
-bytecode-verifier = { path = "../../bytecode-verifier", version = "0.1.0" }
+bcs = "0.1.2"
 diem-crypto = { path = "../../../crypto/crypto", version = "0.1.0" }
 diem-types = { path = "../../../types", version = "0.1.0" }
 diem-workspace-hack = { path = "../../../common/workspace-hack", version = "0.1.0" }
-bcs = "0.1.2"
-vm = { path = "../../vm", version = "0.1.0" }
-diem-framework = { path = "../../diem-framework",  version = "0.1.0" }
```
Comment (mengxu-fb) [inline language/diem-framework/compiled/Cargo.toml:20]: This is the key change I would like to address, dropping the dependency from `compiled-stdlib` to `stdlib` (now `diem-framework`.

The reason is because of the dependency chain:

- ir-compiler
- compiled-stdlib
- diem-framework
- move-lang
- move-prover
- <all-other-kinds-of-stuff>

Therefore, whenever I change the IR-compiler, I need to make sure that all other components compile as well... Which becomes very hard when working on the lifting Vector functions into bytecode instructions PR.

Removing this dependency here will ease the situation a bit. Plus, to me it also makes sense logically as why loading bytes from the disk should require all other components in the repo...

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A move-package 0.1.0 (workspace, path 'language/move-package')
    * features: default
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A move-package 0.1.0 (workspace, path 'language/move-package')
    * features: default
```

File: language/diem-framework/compiled/Cargo.toml (hunk 1 above)
Comment (mengxu-fb) [reply language/diem-framework/compiled/Cargo.toml:20]: No, it is the `compiler/Cargo.toml`

File: common/workspace-hack/Cargo.toml [hunk 2]
```diff
@@ -25,6 +25,7 @@ futures-channel = { version = "0.3.12", features = ["alloc", "default", "futures
 futures-core = { version = "0.3.12", features = ["alloc", "default", "std"] }
 futures-io = { version = "0.3.12", features = ["default", "std"] }
 futures-sink = { version = "0.3.12", features = ["alloc", "default", "std"] }
+futures-task = { version = "0.3.12", default-features = false, features = ["alloc", "once_cell", "std"] }
```
[the remark this answers is not in the document]
Comment (mengxu-fb) [reply common/workspace-hack/Cargo.toml:28]: This is auto-generated

File: language/move-package/src/package.rs [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    name: String,
+    modules_with_bytes: Vec<(Vec<u8>, CompiledModule)>,
+}
+
+#[derive(Clone)]
+pub struct VerifiedPackage {
+    name: String,
+    modules_with_bytes: Vec<(Vec<u8>, CompiledModule)>,
+    dependencies: Vec<VerifiedPackage>,
+}
```
[the remark this answers is not in the document]
Comment (mengxu-fb) [reply language/move-package/src/package.rs:26]: Thanks for the comments Victor!

> packages could form a dependency graph that is not a tree.

I tend to think this is not the case (although in theory it could be), and the reason why I tend to think that packages also form a tree instead of a graph is based on the property that there should be no cyclic dependencies among the modules, and hence, if the packages are grouped reasonably, there should be have no cyclic dependencies among the packages too.

> You want the client to be able to use A with other packages, potentially ones you don't control, then there is no way you can be sure the client uses the version of B owned by A.

I totally agree with you. I have some thoughts on this and the arguments pretty much depend on how the modules are deployed (or rather, published).

- Given the current model, the modules on-chain is the source of truth, and there is no guarantee that what you put in the `dependencies` is the same as the modules on-chain. In that sense, yes, it is a false sense of verified. However, I would argue that the property of having the same on-chain modules it not the problem addressed by a `VerifiedPackage`, it should be enforced by whomever has the access of on-chain modules, and currently, the VM.
  - As a further extension, if we were to have package publishing (e.g., multiple modules in one transaction), the VM could check the dependencies declared in a `VerifiedPackage` against all on-chain modules, and reject if
    - (strict mode) modules mismatch in any byte
    - (compatibility mode) modules that differ do not pass the compatibility check.
- Image there could be an off-chain use case (although that is not the use case I intended), then there is no source-of-truth of modules. Every module that you put in the `VerifiedPackage` is the source of truth and the package is guaranteed to be self-contained.

File: language/diem-framework/compiled/Cargo.toml (hunk 1 above)
Comment (vgao1996) [reply language/diem-framework/compiled/Cargo.toml:20]: Oh I see. I confused myself.

However I'm not sure your suggested change will solve your problem. Note that `language/compiler` and `language/compiler/ir-to-bytecode` are different crates. 
graphviz (5)
When you said you changed the IR compiler, you were probably referring to `ir-to-bytecode`, which is used by the bunch of stuff you mentioned. If I'm understanding correctly, removing the dependency from `compiled-stdlib` to `diem-framework` will only free `language/compiler` from having to depend on a bunch of things but not reduce time of rebuilds triggered by a change in `ir-to-bytecode`.

With that being said I'm not against this change at all!

Comment (mengxu-fb) [reply language/diem-framework/compiled/Cargo.toml:20]: Sorry, it is my bad that I did not specify clearly what the IR compiler is. Thanks for the clarification!

> removing the dependency from compiled-stdlib to diem-framework will only free language/compiler from having to depend on a bunch of things but not reduce time of rebuilds triggered by a change in ir-to-bytecode.

Totally!

File: language/move-package/src/package.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (vgao1996) [reply language/move-package/src/package.rs:26]: > a tree instead of a graph is based on the property that there should be no cyclic dependencies among the modules
Sorry I didn't make this clear. By "not a tree" I meant something like a directed acyclic graph:
graphviz (6)

> the arguments pretty much depend on how the modules are deployed (or rather, published).
I could be wrong, but I feel like concepts like "on-chain" or "module publishing" are orthogonal in this context. Let me explain what I mean by this.

I wasn't really complaining about dependencies of a (verified) package not matching those on-chain. In fact I think it's a different issue and totally agree with you that it's not something the package system should worry about. I also don't see this as a matter of defining the one and only source of truth.

All I was saying was that I didn't feel like `VerifiedModule` was the model clients would want and it would likely create more problems than it would solve. IMO the biggest issue is that `VerifiedModule` is trying to prove an invariant that's broader than its own scope. 

Let's first clarify the property a typical client wants to uphold:
  - "for a given universe of modules, make sure they verify and link"

Here the property applies to a bigger universe, which could contain modules from various sources, for example, different packages or even modules created on the fly. Now trying to impose the property on a single package will result in the package carrying a bigger universe, or even make it the universe itself, which kinda defeats the purpose of having packages in the first place. Here's a metaphor: all I need is a wheel, but I have to pay for a full car just to demonstrate that the wheel works, even when I just want to install that wheel on another car I already have. 

In other words, imagine what a taking the verified package as parameters:
- If it's `&[VerifiedPackage]`, then how do you deal with `[A -> B, B', C -> B'']`, where `B, B', B''` are all compatible but semantically different, or even worse, where `B, B', B''` are in fact incompatible? In this case the per package verification is not only redundant but could create additional confusion.
- If you force clients to take only one `&VerifiedPackage`, then it's the universe, not a package.

Also consider the point @Tim made in the meeting: packaging is a build system concept, not a language one. I think its role is to provide an easier way to refer to or move things as a bundle and it shouldn't care about syntactic or semantic correctness -- that's the job of the client/consumer.

**Comment by davidiw** [comment]
Old PR, closing out, feel free to re-open.
